src/parse/lex.rs

```diff
@@ -1,22 +1,319 @@
-use super::token::{Token, TokenSpan};
+use super::token::{Token, TokenSpan, Lit};
+use std::str::Chars;
 use super::Span;
+use std::iter::{Iterator};
 
+///A lexer with its associated query, a char iterator, and
+///positions (last, current, next)
 pub struct Lexer<'a> {
     query: &'a str,
+    chs : Chars<'a>,
+    last: Option<char>,
+    last_pos: Option<usize>,
     curr: Option<char>,
-    curr_pos: Option<u32>,
+    curr_pos: Option<usize>,
+    next: Option<char>,
+    span_start: Option<usize>
 }
 
 impl<'a> Lexer<'a> {
+
+    ///new lexer, everything set to None
     pub fn from_query<'b>(query: &'b str) -> Lexer<'b> {
-        Lexer {
+        let mut lex = Lexer {
             query: query,
             curr: None,
-            curr_pos: None
+            curr_pos: Some(0),
+            last: None,
+            last_pos: Some(0),
+            next: None,
+            span_start: Some(0),
+            chs : query.chars()
+        };
+        lex.dbump();
+        lex
+    }
+
+    ///bumper function advances to the next char
+    fn bump(&mut self){
+
+        //advance all pointers to the next char
+        self.last = self.curr;
+        self.curr = self.next;
+        self.next = self.chs.next();
+
+
+        //TODO: position-management
+        // advance last_pos to position of current char
+        self.last_pos = self.curr_pos;
+
+        // next position is current position plus the utf8 length
+        // of the next character
+        match self.next {
+            Some(c) => {
+                self.curr_pos = Some(self.curr_pos.unwrap()
+                    + c.len_utf8())
+            }
+            _ => {}
+        };
+
+    }
+
+    ///double bump
+    fn dbump(&mut self) {
+        self.bump();
+        self.bump();
+    }
+
+    ///scan each new WORD from the query string
+    fn scan_words(&mut self) -> String {
+        let mut s = String::new();
+        //loop until the end of a word (only letters and _)
+        loop {
+            //take current char
+            match self.curr.unwrap_or(' ') {
+                c @ 'a' ... 'z' |
+                c @ 'A' ... 'Z' |
+                c @ '0' ... '9' |
+                c @ '_' => {
+                    //push letter into return string
+                    s.push(c);
+                },
+                //if no letter, then stop loop
+                _ => break
+            }
+            //next char
+            self.bump();
+        }
+        s
+    }
+
+    ///scan each new NUMBER from the query string
+    fn scan_nums(&mut self) -> String {
+        let mut s = String::new();
+        let mut dot = false;
+        loop {
+            match self.curr.unwrap_or(' ') {
+                c @ '0' ... '9' => {
+                    s.push(c);
+                }
+                c @ '.' => {
+                    if dot {
+                        println!("This is not a good number, Sir!");
```

Comment:
You should not have any user interaction in your `parse` module :P 
But I'm sure this is just temporary, Madam.